diesel/src/sqlite/connection/stmt.rs

```diff
@@ -189,88 +188,113 @@ impl Drop for Statement {
 
 struct BoundStatement<'stmt, 'query> {
     statement: MaybeCached<'stmt, Statement>,
-    // we need to store the query here to ensure noone does
-    // drop it till the end ot the statement
-    // We use a boxed queryfragment here just to erase the
-    // generic type, we use ManuallyDrop to communicate
-    // that this is a shared buffer
-    query: ManuallyDrop<Box<dyn QueryFragment<Sqlite> + 'query>>,
     // we need to store any owned bind values speratly, as they are not
-    // contained in the query itself. We use ManuallyDrop to
-    // communicate that this is a shared buffer
-    binds_to_free: ManuallyDrop<Vec<(i32, Option<SqliteBindValue<'static>>)>>,
```

Comment:
As far as I understand the current state of rusts unsafe code guide lines removing the `ManuallyDrop` here may cause be undefined behaviour. The problem here is that internally this vector stores either a `String` or a `Vec<u8>` which both internally use a `Unique` pointer. As the name suggests this pointer type assumes unique access to the allocated memory.  As soon as we bind the buffer to the sqlite statement we violate this contract. According to [this](https://users.rust-lang.org/t/code-review-for-unsafe-code-in-diesel/66798/8) is one possibility to communicate to the compiler that it cannot assume this property. (The same argument applies to the `query` field above)

Reply:
This is fine as long as the corresponding pointer is never used by anything while the other references exist : the unique property is used to enable noalias llvm optimizations, which won't enter into play as long as the other reference is never deferenced.
There are similar issues describing this and what would be actually valid within https://github.com/Kimundi/owning-ref-rs/issues/49.
It can be seen that the issue arises from the code accessing the same memory location through pointers that have noalias.
Note that in this PR, the box is never dereferenced after binds are collected and until they are dropped - which is the exact same that would happen when writing normal code.

Reply:
I had this in mind when writing it but it's definitely something I should have commented on as it's definitely super not obvious 😄

Reply:
As far as I understand the owning-ref issue and the [linked unsafe code guideline](https://github.com/rust-lang/unsafe-code-guidelines/issues/194#issuecomment-520934222) issue the topic is:
a) Not decided yet
b) `Box<T>` contains a `Unique<T>` pointer which gets translated as `noalias` so yes technically as long as we don't access the pointer till we resolved the aliasing we should be fine. But that seems a property that is really hard to verify. Especially around drop. For example the compiler already notices that this field is "unused" so it may want to decide to just not store the corresponding struct field till `BoundStatement` is used. `ManuallyDrop` resolves the `Drop` issue, as we then can exactly tell the compiler to perform the drop at a specific location. It does not really resolve any other potential use through the `Unique` ptr there. It seems like the only way that is currently accepted as definitively not causing UB is to use [`Box::into_raw`](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.into_raw) here to move out the pointer from the inner `Unique<T>` type into a normal pointer. This could be stored till the any aliasing is over (so till the `Drop` impl) and then reconstruct the Box via `Box::from_raw`. (Same reasoning for `String`/`Vec` as they are internally similar, but there we need to store the length as well).

Especially see [that linked playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018) (notice the difference between release and debug build :scream: )

Note: This discussion likely applies to the [mysql bind implementation](https://github.com/diesel-rs/diesel/blob/master/diesel/src/mysql/connection/bind.rs) as well. :disappointed: 

(I should note that the more I read about aliasing, the more I come horrified how to ever get that right in the context of non-trivial unsafe code like that one :scream: )

Reply:
(playground is link broken)

> the more I read about aliasing, the more I come horrified how to ever get that right in the context of non-trivial unsafe code like that one 😱

Yeah I didn't realize how hard `unsafe` actually was until I had a look at the open issues in owning-ref either 😅

> that seems a property that is really hard to verify. Especially around drop. For example the compiler already notices that this field is "unused" so it may want to decide to just not store the corresponding struct field till `BoundStatement` is used.

AFAIK `Drop`'s behavior is always specified - rust wouldn't call `drop` on the `BoundStatement` without having set its `_query` field (nor would it ever do that with any other function).
I'm not sure I understand what scenario you have in mind here.

Considering we don't need to use the query and only keep it here to prevent it from being destructed, I believe it's easier to guarantee that a field is never used than it is to guarantee that we don't forget to drop stuff on error/ok/panic. (e.g. in the previous implementation, if anything happens to _panic_ within `for (bind_idx, (bind, tpe)) in (1..).zip(binds.into_iter().zip(metadata))`, `query` and `binds_to_free` are memory-leaked)
(Then again I should definitely have added comments specifying that this field should never be used because of noalias requirements.)

To the extreme, if we want to compile-enforce that the field is never used, we can
```rust
mod never_use {
    pub struct NeverUse<T> {
        _private: T,
    }
    impl<T> NeverUse<T> {
        fn new(inner: T) -> Self {
            Self { _private: inner }
        }
    }
}
pub use never_use::NeverUse;
```

I'd feel much less confident guaranteeing correctness of rust code that manages its allocations itself with raw pointers manipulations than guaranteeing that the pointer is never dereferenced until dropped, and that it is never dropped before the relevant binds are re-bound, using RAII.

Reply:
> a) Not decided yet

Not sure I understand what you mean. It seems that regardless of whether they decide that [this code](https://github.com/rust-lang/unsafe-code-guidelines/issues/194#issuecomment-520934222) is UB or not, as long as we *don't use* the "unique" at the same time as a pointer derived from it, we're fine?

Reply:
[Fixed playground link](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=482426bd66160ed084bfbe9a4f896d8d)

> I'd feel much less confident guaranteeing correctness of rust code that manages its allocations itself with raw pointers manipulations than guaranteeing that the Box is never dereferenced until dropped, and that it is never dropped before the relevant binds are re-bound, using RAII.

I feel exactly the other way around. Handling memory allocations is something I'm quite confident about, while I fear that the compiler may decide that this are not used and therefore   removes something. Maybe I'm just trying to be too perfect here...

Otherwise I've played a bit with an modified implementation that stores raw pointers. I've got a working version that:
* Uses drop as single "cleanup" point
* Only contains a minimal trivial amount of unsafe ( no transmute, no pointer case, just `Box::from_raw` and `NonNull::new_unchecked`)

I personally find that version quite readable and surprisingly simple compared to any other version yet. I will add some more comments tomorrow and push that as new commit so we can discuss that version.
Generally I would prefer to error on the "safe" side here, so if something is not that obvious from the unsafe code guidelines better use something simpler...

Reply:
As promised I pushed https://github.com/weiznich/diesel/commit/28409cca76ddc9e1702db8421c50e3c11a6ba6be which uses raw pointers.

Reply:
(The PR is updated as well)